core/drivers/bcm_sotp.c

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2019 Broadcom.
+ */
+
+#include <assert.h>
+#include <drivers/bcm_sotp.h>
+#include <initcall.h>
+#include <io.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+
+#define SOTP_PROG_CONTROL			0x0
+#define SOTP_PROG_CONTROL__OTP_CPU_MODE_EN	(1 << 15)
+#define SOTP_PROG_CONTROL__OTP_DISABLE_ECC	(1 << 9)
+#define SOTP_ADDR__OTP_ROW_ADDR_R		6
+
+#define SOTP_ADDR				0xc
+
+#define SOTP_CTRL_0				0x10
+#define SOTP_CTRL_0__START			(1 << 0)
+#define SOTP_READ				(0 << 1)
+
+#define SOTP_STAT_0				0x18
+#define SOTP_STATUS_0__FDONE			(1 << 3)
+
+#define SOTP_STATUS_1				0x1c
+#define SOTP_STATUS_1__CMD_DONE			(1 << 1)
+#define SOTP_STATUS_1__ECC_DET			(1 << 17)
+
+#define SOTP_RDDATA_0				0x20
+#define SOTP_RDDATA_1				0x24
+#define SOTP_ADDR_MASK				0x3ff
+
+#define SOTP_ECC_ERR_DETECT			0x8000000000000000
+
+#define MAX_TIME_OUT				30000
+
+struct bcm_sotp_dev *bcm_sotp;
```

Comment:
Same comment as in `bcm_hwrng.c`:
Instead of lines 39-43 I'd just use a static vaddr_t sotp_base;
It simplifies the code and avoids a call to malloc().